### PR TITLE
test: Add test for global rate limiting with load balancing

### PR DESCRIPTION
**Pull request by philipgough**
Draft PR to replicate the behaviour/bug seen in #208

### CODE REVIEW

File: functional_test.go [hunk 1]
```diff
@@ -859,6 +864,62 @@ func TestGlobalRateLimits(t *testing.T) {
 	})
 }
 
+func TestGlobalRateLimitsWithLoadBalancing(t *testing.T) {
```
Comment (miparnisari) [inline functional_test.go:867]: As a non-maintainer of this repo, it looks like a code smell to me that the easiest way to assert this behavior is through a functional test and not a unit test. 🤔

**Comment by thrawn01** [comment]
What are we testing here? GPRC's load balancing or the ability for gubernator to distribute the hits to the rest of the cluster?

In reference to unit testing vs functional. If we can have a unit test which tests the *ahem* functionality without violating public interfaces, then we can do so. Any test which violates the public interface is suspect as it limits your ability to refactor in the future. Functional Tests > Unit Tests. 

You can read my thoughts on this here https://wippler.dev/posts/Testing-private-methods

**Comment by thrawn01** [comment]
ah, I didn't realize this PR was related to #208 👍

**Comment by philipgough** [comment]
Yep @thrawn01 - I just created this to reproduce the exact issue I reported in #208 for clarity!

File: functional_test.go [hunk 2]
```diff
@@ -859,6 +864,62 @@ func TestGlobalRateLimits(t *testing.T) {
...
+	}
+
+	address := fmt.Sprintf("static:///%s,%s", owner, peer)
+	conn, err := grpc.DialContext(context.Background(), address, dialOpts...)
+	require.NoError(t, err)
+
+	client := guber.NewV1Client(conn)
+
+	sendHit := func(status guber.Status, assertion func(resp *guber.RateLimitResp), i int) string {
+		ctx, cancel := context.WithTimeout(context.Background(), clock.Hour*5)
```
Comment (miparnisari) [inline functional_test.go:885]: this timeout is waaaay too high. Github actions will not run for more than 10 minutes by default

File: functional_test.go [hunk 3]
```diff
@@ -859,6 +864,62 @@ func TestGlobalRateLimits(t *testing.T) {
...
+		assert.Equal(t, status, gotResp.GetStatus(), i)
+
+		if assertion != nil {
+			assertion(gotResp)
+		}
+
+		return gotResp.GetMetadata()["owner"]
+	}
+
+	// Send two hits that should be processed by the owner and the peer and deplete the limit
```
Comment (miparnisari) [inline functional_test.go:912]: > should be processed by the owner

are you asserting this anywhere? I don't see it

File: functional_test.go [hunk 4]
```diff
@@ -859,6 +864,62 @@ func TestGlobalRateLimits(t *testing.T) {
...
+	assert.NotEqual(t, owner, peer)
+
+	dialOpts := []grpc.DialOption{
+		grpc.WithResolvers(newStaticBuilder()),
+		grpc.WithTransportCredentials(insecure.NewCredentials()),
+		grpc.WithDefaultServiceConfig(`{"loadBalancingConfig": [{"round_robin":{}}]}`),
+	}
+
+	address := fmt.Sprintf("static:///%s,%s", owner, peer)
+	conn, err := grpc.DialContext(context.Background(), address, dialOpts...)
```
Comment (miparnisari) [inline functional_test.go:879]: can you clarify what this is doing? it's not clear to me

File: functional_test.go [hunk 5]
```diff
@@ -859,6 +864,62 @@ func TestGlobalRateLimits(t *testing.T) {
...
+		}
+
+		return gotResp.GetMetadata()["owner"]
+	}
+
+	// Send two hits that should be processed by the owner and the peer and deplete the limit
+	sendHit(guber.Status_UNDER_LIMIT, nil, 1)
+	sendHit(guber.Status_UNDER_LIMIT, nil, 2)
+	// sleep to ensure the async forward has occurred and state should be shared
+	time.Sleep(time.Second * 5)
```
Comment (miparnisari) [inline functional_test.go:916]: I think (not sure) you can control this via setting `globalBatchTimeout = 1` and `globalSyncWait = 1ms`, so you don't need to wait so long here

**Comment by miparnisari** [comment]
@thrawn01 

> What are we testing here? GPRC's load balancing or the ability for gubernator to distribute the hits to the rest of the cluster?

I didn't write the test but the goal should be to assert that the gossiping of the rate limiting from owner to peers is correct in that the `status` field is correct.

> Functional Tests > Unit Tests.

I don't know what you meant by this but if a bug happens in a confined place (which is the case here), you should not have to write a functional test to demonstrate it. The test should be laser-focused on what is broken. Functional tests cover a lot more code.

File: functional_test.go (hunk 2 above)
Comment (philipgough) [reply functional_test.go:885]: just to be clear, this test wasnt ready to be merged as is, hence why it was in draft :) - it was simply to reproduce the culprit behaviour that was reported in the bug.

File: functional_test.go (hunk 4 above)
Comment (philipgough) [reply functional_test.go:879]: this is creating a static pool of servers to communicate with and creating the connection

Comment (miparnisari) [reply functional_test.go:879]: the connection is to which server?

Comment (philipgough) [reply functional_test.go:879]: between the owner and the peer

File: functional_test.go (hunk 3 above)
Comment (philipgough) [reply functional_test.go:912]: no in this case im relying on the grpc load balancer we passed

Comment (philipgough) [reply functional_test.go:912]: but we can and should check in order to harden the test

**Comment by thrawn01** [comment]
@miparnisari re functional testing.

I understand that most developers are taught to write unit tests and thus are surprised when they encounter a project which is mostly functional tests. This is a common issue when we on boarding new developers at Mailgun. Please have confidence when making changes to `algorithms.go` that those functions are well tested. A coverage run shows 81% of statements in `algorithms.go` are tested with  1,195 calls during the functional test suite run. I'm not claiming the current testing level is adequate, (we could improve coverage on `Behavior_DURATION_IS_GREGORIAN` for instance) only that functional testing allows developers to freely change the internals of a system while having high confidence they didn't break functionality.

File: functional_test.go (hunk 5 above)
Comment (Baliedge) [reply functional_test.go:916]: Or, use `github.com/mailgun/holster/v4/clock` to freeze or advance system time during the test.

**Comment by Baliedge** [comment]
@philipgough Thank you very much for your contribution.  I was able to reproduce the error using your test.  And since the PR has a merge conflict and it's been some time since created, in a separate branch I updated from `master` and retried the test to see if recent changes from https://github.com/mailgun/gubernator/pull/219 had addressed your issue.

I found some improvement.  Sometimes the test passes:
```
=== RUN   TestGlobalRateLimitsWithLoadBalancing
--- PASS: TestGlobalRateLimitsWithLoadBalancing (5.00s)
PASS
ok      github.com/mailgun/gubernator/v2        6.074s
```

Other times I get 2 failed assertions:
```
=== RUN   TestGlobalRateLimitsWithLoadBalancing
    functional_test.go:1060:
                Error Trace:    /Users/spoulson/src/gubernator2/functional_test.go:1060
                                                        /Users/spoulson/src/gubernator2/functional_test.go:1070
                Error:          Not equal:
                                expected: 0
                                actual  : 1
                Test:           TestGlobalRateLimitsWithLoadBalancing
                Messages:       1
    functional_test.go:1060:
                Error Trace:    /Users/spoulson/src/gubernator2/functional_test.go:1060
                                                        /Users/spoulson/src/gubernator2/functional_test.go:1071
                Error:          Not equal:
                                expected: 0
                                actual  : 1
                Test:           TestGlobalRateLimitsWithLoadBalancing
                Messages:       2
--- FAIL: TestGlobalRateLimitsWithLoadBalancing (5.01s)
```

But this is better than the original code that failed 5 assertions.  We're on the right path.

You can refer to my branch to cherry-pick my commits if needed: https://github.com/mailgun/gubernator/compare/master...Baliedge/global-lb

**Comment by Baliedge** [comment]
Closed in favor of https://github.com/mailgun/gubernator/pull/224.
I applied the edits I mentioned above and some other adjustments.  Test passes every time.

**Pull request closed without merging**